### ISSUE TITLE
[BUG] "Spike no more" feature was removed

### ISSUE THREAD

**Issue by cyr0930** [NVIDIA/Megatron-LM#3945]
https://github.com/NVIDIA/Megatron-LM/pull/2047
this PR removed this feature (https://github.com/NVIDIA/Megatron-LM/blob/v0.15.3/megatron/training/arguments.py#L2302)

also see "default_skip_embedding_weight_decay"

**Comment by DhineshPonnarasan** [comment]
Hi @cyr0930 ,
Shall I work on this?

**Comment by cyr0930** [comment]
> Hi [@cyr0930](https://github.com/cyr0930) , Shall I work on this?

of course. thanks in advance :)

**Comment by Phlip79** [comment]
What is your use-case for this feature @cyr0930?

**Comment by cyr0930** [comment]
@Phlip79 I will use this feature for training stabilizer.
But I am not sure it really works in practice.
I think it can be removed if that decision is intentional.

**Comment by dimapihtar** [comment]
Hi @cyr0930, is this issue still relevant? Thanks!

**Comment by cyr0930** [comment]
@dimapihtar Hi!
I found some researchers think applying wd on embedding is not right (https://github.com/karpathy/minGPT/pull/24#issuecomment-679316025), but I am not sure as some research experiments support that argument and some don't.
And the fact that decaying embedding vectors is the global default may support "it doesn't matter that much".
So I think it depends on that this feature is worth occupying one argument or not.
I don't believe it is necessary. It looks optional.

**Comment by Connor-XY** [comment]
Hello @cyr0930, thanks for pointing out the removed behavior and for following up with the embedding-weight-decay context.

I traced this back to #2047. The old `--embedding-init-method-std` path effectively bundled two behaviors: a separate embedding init std and skipping weight decay on embedding weights. After the optimizer interface refactor, I do not see a direct replacement for that combined behavior in the current tree.

Given your latest note that this is optional and may be a design choice, I’m treating this as an optimizer policy/design decision rather than an active correctness bug. I’m routing this to @Phlip79 from the MCore oncall rotation, who had already engaged on the thread, to decide whether to restore this as an explicit parameter-group option or close it as intentionally removed.

**Comment by cspades** [comment]
cc @jstjohn and Tyler Shimko

@deepakn94 Why did we remove this feature? This is something John and Tyler worked on and we saw very nice statistics in loss spikes while training Evo2.

**Comment by jstjohn** [comment]
This should all still be supported in megatron lm, but how you do it changed. As @Connor-XY pointed out, there are a few things you might want to do with spike-no-more. The first is what the paper calls out, that you should set the embedding stdev to 1.0. That is something you can do in your megatron config (https://github.com/NVIDIA/Megatron-LM/blob/3f6a2ed/megatron/core/transformer/transformer_config.py#L350). By itself that should get you the key part of the spike-no-more paper by setting that to `embedding_init_method_std=1.0`. The other part you will probably want to do is make sure that embeddings are not tied to the output layer, since 1.0 is a pretty aggressive init (`share_embeddings_and_output_weights=False`). 

Next, I don't think they did this in the paper, and if you track the param norm during training you'll probably notice that the embedding weights never become too small even if you do not do this, but you could set weight decay to 0 for the embedding layer so that it never shrinks to 0 even in a very long training run.

If you want to set weight decay to 0 for the embedding layer, you'll need to modify the training script for your run, or maybe request that an option for this be added to `training.py` if you are using that script as is and want a single flag to toggle this. With the new megatron api for modifying settings for param groups you can do something like the following with megatron bridge making use of the new megatron core api:
https://github.com/NVIDIA-BioNeMo/bionemo-framework/blob/main/bionemo-recipes/recipes/evo2_megatron/src/bionemo/evo2/models/evo2_provider.py#L119-L120 where we add regex matches to embedding layers to the set that have `wd_mult=0`. You can see similar settings in megatron training scripts  around https://github.com/NVIDIA/Megatron-LM/blob/b57449928aafaaec7a36e8325d83aacf05e91bd4/megatron/core/optimizer/__init__.py#L117, which gets used in your train script here: https://github.com/NVIDIA/Megatron-LM/blob/b57449928aafaaec7a36e8325d83aacf05e91bd4/megatron/training/training.py#L1894. You could append another match/override for setting wd=0 for matches to just embedding layers around that line of your train script if you wanted to disable wd on embeddings. Megatron will take the union of overrides, so you can just append another override+setting to that default overrides list.

**Comment by cspades** [comment]
Just checking in, has this issue been resolved? @cyr0930 @DhineshPonnarasan

**Comment by DhineshPonnarasan** [comment]
> Just checking in, has this issue been resolved? [@cyr0930](https://github.com/cyr0930) [@DhineshPonnarasan](https://github.com/DhineshPonnarasan)

Hi @cspades , I investigated the current main branch after noticing that my previous PR (#4012) was closed.

From what I can see, the issue still appears to be present on the current main. The behavior described for skipping weight decay on embedding parameters when embedding_init_method_std is set doesn't appear to be implemented in the current optimizer flow.

Since the codebase has evolved since #4012, Shall I prepare a fresh implementation based on the current main instead of trying to revive the old branch. And I'll open a new PR once it's ready.

**Comment by cspades** [comment]
@DhineshPonnarasan I think the PR being closed was somewhat of a misunderstanding, basically @jstjohn 's suggestion above is what you would want to PR into MLM:

1. `embedding_init_method_std=1.0`
2. `share_embeddings_and_output_weights=False`
3. `wd_mult=0`

For (3), is that already the case? https://github.com/NVIDIA/Megatron-LM/blob/c1560d9481dae7c5243623457f5a1788d411cc6b/megatron/core/optimizer/__init__.py#L119

I think the goal is to enable the above 3 `training.py` arguments / configs instead of tying their logic together, if you can check if there's still an issue, and if there is, modify your PR to be lightweight we can review!

**Comment by DhineshPonnarasan** [comment]
> [@DhineshPonnarasan](https://github.com/DhineshPonnarasan) I think the PR being closed was somewhat of a misunderstanding, basically [@jstjohn](https://github.com/jstjohn) 's suggestion above is what you would want to PR into MLM:
> 
> 1. `embedding_init_method_std=1.0`
> 2. `share_embeddings_and_output_weights=False`
> 3. `wd_mult=0`
> 
> For (3), is that already the case?
> 
> [Megatron-LM/megatron/core/optimizer/__init__.py](https://github.com/NVIDIA/Megatron-LM/blob/c1560d9481dae7c5243623457f5a1788d411cc6b/megatron/core/optimizer/__init__.py#L119)
> 
> Line 119 in [c1560d9](/NVIDIA/Megatron-LM/commit/c1560d9481dae7c5243623457f5a1788d411cc6b)
> 
>  config_overrides[param_wd_mult_key] = ParamGroupOverride(wd_mult=0.0) 
> I think the goal is to enable the above 3 `training.py` arguments / configs instead of tying their logic together, if you can check if there's still an issue, and if there is, modify your PR to be lightweight we can review!

---

@cspades , Thanks for the clarification. I rechecked the current upstream main and verified that the embedding wd_mult=0 behavior is already implemented via the existing ParamGroupOverride path when embedding_init_method_std is set. Given that, I don't see the original issue reproducing on the current codebase. Am I understanding correctly that the remaining goal is to make wd_mult=0 independently configurable rather than coupled to embedding_init_method_std?

**Comment by cspades** [comment]
I think if we can replicate spike-no-more and there's no reason to make `wd_mult=0` independently configurable, then we can close this issue!

**Comment by DhineshPonnarasan** [comment]
> I think if we can replicate spike-no-more and there's no reason to make `wd_mult=0` independently configurable, then we can close this issue!

---

From what I found, the current optimizer flow still doesn't reproduce the original Spike No More behavior. Although both TransformerConfig and TrainingConfig document that setting embedding_init_method_std should disable weight decay on embedding weights, the optimizer configuration only applies wd_mult=0 to bias and 1D parameters. Embedding weight matrices still receive the default wd_mult=1.0.

It looks like the behavior was documented but isn't currently wired into the optimizer path.

**Comment by cspades** [comment]
> Am I understanding correctly that the remaining goal is to make wd_mult=0 independently configurable rather than coupled to embedding_init_method_std?

Okay, I think this could make sense. Can you open the PR for an argument to control `wd_mult` directly possibly without being linked to that embedding_init_method_std? I'll have some people (who understand this better) take a look.

Default behavior should stay the same, i.e. preserve the coupling, however you want to make that happen.

**Comment by DhineshPonnarasan** [comment]
> > Am I understanding correctly that the remaining goal is to make wd_mult=0 independently configurable rather than coupled to embedding_init_method_std?
> 
> Okay, I think this could make sense. Can you open the PR for an argument to control `wd_mult` directly possibly without being linked to that embedding_init_method_std? I'll have some people (who understand this better) take a look.
> 
> Default behavior should stay the same, i.e. preserve the coupling, however you want to make that happen.

---


@cspades  thanks! I'll prepare a lightweight PR that adds a dedicated configuration argument to control embedding wd_mult independently of embedding_init_method_std, following the current optimizer configuration flow. I'll keep the change minimal and let the optimizer experts weigh in on the API during review.

**Comment by DhineshPonnarasan** [comment]
@cspades I have a question. Before I implement it, I just wanted to clarify the API. When you mentioned "an argument to control wd_mult directly", were you envisioning:
- a boolean flag (e.g. --skip-embedding-weight-decay) that internally sets wd_mult=0.0 or a float-valued argument (e.g. --embedding-wd-mult) that lets users set the multiplier explicitly?

I can implement either; I just want to make sure the public API matches what you had in mind.
